In [68]:
# ============================================================
# MULTIVARIATE ML FORECASTING
# LOAD MULTIPLE CBS TABLES
# ============================================================

# ------------------------------------------------------------
# Imports
# ------------------------------------------------------------

import numpy as np
import pandas as pd

from src.config import DIR_DB_SILVER
from src.utils.m_query_database import f_query_database
from sklearn.ensemble import GradientBoostingRegressor

# ============================================================
# LOAD ABSENTEEISM TABLE
# 80072NED
# ============================================================

query_abs = '''
SELECT *
FROM "80072ned_silver"
'''

df_abs = f_query_database(
    db_path=DIR_DB_SILVER,
    query=query_abs,
    return_type="pandas"
)

print("Absenteeism shape:")
print(df_abs.shape)

# ============================================================
# LOAD LABOR VOLUME TABLE
# 85920NED
# ============================================================

query_labor = '''
SELECT *
FROM "85920ned_silver"
'''

df_labor = f_query_database(
    db_path=DIR_DB_SILVER,
    query=query_labor,
    return_type="pandas"
)

print("\nLabor volume shape:")
print(df_labor.shape)

# ============================================================
# LOAD COMPENSATION TABLE
# 85917NED
# ============================================================

query_comp = '''
SELECT *
FROM "85917ned_silver"
'''

df_comp = f_query_database(
    db_path=DIR_DB_SILVER,
    query=query_comp,
    return_type="pandas"
)

print("\nCompensation shape:")
print(df_comp.shape)

print("\n===== LABOR TABLE COLUMNS =====")
print(df_labor.columns.tolist())

print("\n===== COMPENSATION TABLE COLUMNS =====")
print(df_comp.columns.tolist())

display(df_labor.head())
display(df_comp.head())

✅ Query returned 5850 rows from silver_data.db as pandas DataFrame
✅ Query returned 7440 rows from silver_data.db as pandas DataFrame
✅ Query returned 2464 rows from silver_data.db as pandas DataFrame


Absenteeism shape:
(5850, 13)

Labor volume shape:
(7440, 26)

Compensation shape:
(2464, 44)

===== LABOR TABLE COLUMNS =====
['silver_id', 'bronze_pk', 'ID', 'TypeWerkenden', 'BedrijfstakkenBranchesSBI2008', 'Perioden', 'WerkzamePersonen_1', 'WerkzamePersonenSeizoengecorrigeerd_2', 'Banen_3', 'BanenSeizoengecorrigeerd_4', 'GewerkteUren_5', 'GewerkteUrenSeizoengecorrigeerd_6', 'WerkzamePersonen_7', 'Banen_8', 'WerkzamePersonenSeizoengecorrigeerd_9', 'BanenSeizoengecorrigeerd_10', '_source_file', 'BedrijfstakkenBranchesSBI2008_Title', 'BedrijfstakkenBranchesSBI2008_Description', 'BedrijfstakkenBranchesSBI2008_CategoryGroupID', 'Perioden_Title', 'Perioden_Description', 'Perioden_Status', 'TypeWerkenden_Title', 'TypeWerkenden_Description', 'TypeWerkenden_CategoryGroupID']

===== COMPENSATION TABLE COLUMNS =====
['silver_id', 'bronze_pk', 'ID', 'BedrijfstakkenBranchesSBI2008', 'Perioden', 'BeloningVanWerknemers_1', 'BeloningSeizoengecorrigeerd_2', 'Lonen_3', 'LonenSeizoengecorrigeerd_4', 

,silver_id,bronze_pk,ID,TypeWerkenden,BedrijfstakkenBranchesSBI2008,Perioden,WerkzamePersonen_1,WerkzamePersonenSeizoengecorrigeerd_2,Banen_3,BanenSeizoengecorrigeerd_4,...,_source_file,BedrijfstakkenBranchesSBI2008_Title,BedrijfstakkenBranchesSBI2008_Description,BedrijfstakkenBranchesSBI2008_CategoryGroupID,Perioden_Title,Perioden_Description,Perioden_Status,TypeWerkenden_Title,TypeWerkenden_Description,TypeWerkenden_CategoryGroupID
0,1,TypedDataSet.json_0,0,T001413,T001081,1995KW01,7132.0,7215.0,7913.0,8000.0,...,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 1e kwartaal,,Definitief,Totaal,Het totaal van de werknemers en zelfstandigen.,None
1,2,TypedDataSet.json_1,1,T001413,T001081,1995KW02,7273.0,7249.0,8069.0,8043.0,...,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 2e kwartaal,,Definitief,Totaal,Het totaal van de werknemers en zelfstandigen.,None
2,3,TypedDataSet.json_2,2,T001413,T001081,1995KW03,7363.0,7297.0,8167.0,8097.0,...,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 3e kwartaal,,Definitief,Totaal,Het totaal van de werknemers en zelfstandigen.,None
3,4,TypedDataSet.json_3,3,T001413,T001081,1995KW04,7351.0,7358.0,8158.0,8167.0,...,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 4e kwartaal,,Definitief,Totaal,Het totaal van de werknemers en zelfstandigen.,None
4,5,TypedDataSet.json_4,4,T001413,T001081,1995JJ00,7281.0,None,8078.0,None,...,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995,,Definitief,Totaal,Het totaal van de werknemers en zelfstandigen.,None


,silver_id,bronze_pk,ID,BedrijfstakkenBranchesSBI2008,Perioden,BeloningVanWerknemers_1,BeloningSeizoengecorrigeerd_2,Lonen_3,LonenSeizoengecorrigeerd_4,SocialePremiesTLVWerkgevers_5,...,Totaal_30,Mannen_31,Vrouwen_32,_source_file,BedrijfstakkenBranchesSBI2008_Title,BedrijfstakkenBranchesSBI2008_Description,BedrijfstakkenBranchesSBI2008_CategoryGroupID,Perioden_Title,Perioden_Description,Perioden_Status
0,1,TypedDataSet.json_0,0,T001081,1995KW01,38433.0,40204.0,31827.0,33627.0,6606.0,...,None,None,None,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 1e kwartaal,,Definitief
1,2,TypedDataSet.json_1,1,T001081,1995KW02,46140.0,41191.0,39378.0,34562.0,6762.0,...,None,None,None,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 2e kwartaal,,Definitief
2,3,TypedDataSet.json_2,2,T001081,1995KW03,39269.0,41735.0,32724.0,35059.0,6545.0,...,None,None,None,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 3e kwartaal,,Definitief
3,4,TypedDataSet.json_3,3,T001081,1995KW04,41718.0,42475.0,35021.0,35743.0,6697.0,...,None,None,None,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995 4e kwartaal,,Definitief
4,5,TypedDataSet.json_4,4,T001081,1995JJ00,165560.0,None,138950.0,None,26610.0,...,None,None,None,TypedDataSet.json,A-U Alle economische activiteiten,Alle economische activiteiten\r\nDeze categori...,None,1995,,Definitief


In [69]:
# ============================================================
# CREATE TEMPORAL FEATURES
# FOR NEW TABLES
# ============================================================

def prepare_periods(df):

    # --------------------------------------------------------
    # Quarterly only
    # --------------------------------------------------------

    df = (
        df[
            df["Perioden"].str.contains("KW")
        ]
        .copy()
    )

    # --------------------------------------------------------
    # Year
    # --------------------------------------------------------

    df["year"] = pd.to_numeric(
        df["Perioden"].str[:4],
        errors="coerce"
    )

    # --------------------------------------------------------
    # Quarter
    # --------------------------------------------------------

    df["quarter"] = (
        df["Perioden"]
        .str.extract(r"KW(\d{2})")[0]
    )

    df["quarter"] = pd.to_numeric(
        df["quarter"],
        errors="coerce"
    )

    # --------------------------------------------------------
    # Quarter end date
    # --------------------------------------------------------

    df["period_enddate"] = (
        pd.PeriodIndex.from_fields(
            year=df["year"].astype(int),
            quarter=df["quarter"].astype(int),
            freq="Q"
        )
        .to_timestamp(how="end")
        .normalize()
    )

    return df


# ============================================================
# APPLY
# ============================================================

df_labor = prepare_periods(df_labor)

df_comp = prepare_periods(df_comp)

print(df_labor.shape)
print(df_comp.shape)

print(
    df_labor[
        "BedrijfstakkenBranchesSBI2008_Title"
    ].nunique()
)

print(
    df_comp[
        "BedrijfstakkenBranchesSBI2008_Title"
    ].nunique()
)

(5952, 29)
(1984, 47)
16
16


In [70]:
# ============================================================
# UNIQUE SECTORS
# ============================================================

abs_sectors = set(
    df_abs[
        "BedrijfskenmerkenSBI2008_Title"
    ].unique()
)

labor_sectors = set(
    df_labor[
        "BedrijfstakkenBranchesSBI2008_Title"
    ].unique()
)

comp_sectors = set(
    df_comp[
        "BedrijfstakkenBranchesSBI2008_Title"
    ].unique()
)

# ============================================================
# COMMON SECTORS
# ============================================================

common_sectors = (
    abs_sectors
    .intersection(labor_sectors)
    .intersection(comp_sectors)
)

print("\nCOMMON SECTORS:")
print(sorted(common_sectors))

print("\nCOUNT:")
print(len(common_sectors))


COMMON SECTORS:
['A Landbouw, bosbouw en visserij', 'A-U Alle economische activiteiten', 'C Industrie', 'F Bouwnijverheid', 'J Informatie en communicatie', 'K Financiële dienstverlening', 'L Verhuur en handel van onroerend goed', 'O Openbaar bestuur en overheidsdiensten', 'P Onderwijs', 'Q Gezondheids- en welzijnszorg']

COUNT:
10


In [71]:
# ============================================================
# PREPARE ABSENTEEISM TABLE
# ============================================================

# ------------------------------------------------------------
# Keep quarterly observations only
# ------------------------------------------------------------

df_abs = (
    df_abs[
        df_abs["Perioden"].str.contains("KW")
    ]
    .copy()
)

# ------------------------------------------------------------
# Year
# ------------------------------------------------------------

df_abs["year"] = pd.to_numeric(
    df_abs["Perioden"].str[:4],
    errors="coerce"
)

# ------------------------------------------------------------
# Quarter
# ------------------------------------------------------------

df_abs["quarter"] = (
    df_abs["Perioden"]
    .str.extract(r"KW(\d{2})")[0]
)

df_abs["quarter"] = pd.to_numeric(
    df_abs["quarter"],
    errors="coerce"
)

# ------------------------------------------------------------
# Quarter end date
# ------------------------------------------------------------

df_abs["period_enddate"] = (
    pd.PeriodIndex.from_fields(
        year=df_abs["year"].astype(int),
        quarter=df_abs["quarter"].astype(int),
        freq="Q"
    )
    .to_timestamp(how="end")
    .normalize()
)

print(df_abs.shape)

(4680, 16)


In [72]:
# ============================================================
# KEEP ONLY COMMON SECTORS
# ============================================================

common_sectors = list(common_sectors)

# ------------------------------------------------------------
# ABSENTEEISM
# ------------------------------------------------------------

df_abs_filtered = (
    df_abs[
        df_abs[
            "BedrijfskenmerkenSBI2008_Title"
        ].isin(common_sectors)
    ]
    .copy()
)

# ------------------------------------------------------------
# LABOR
# ------------------------------------------------------------

df_labor_filtered = (
    df_labor[
        df_labor[
            "BedrijfstakkenBranchesSBI2008_Title"
        ].isin(common_sectors)
    ]
    .copy()
)

# ------------------------------------------------------------
# COMPENSATION
# ------------------------------------------------------------

df_comp_filtered = (
    df_comp[
        df_comp[
            "BedrijfstakkenBranchesSBI2008_Title"
        ].isin(common_sectors)
    ]
    .copy()
)

print(df_abs_filtered.shape)
print(df_labor_filtered.shape)
print(df_comp_filtered.shape)

(1200, 16)
(3720, 29)
(1240, 47)


In [73]:
# ============================================================
# ABSENTEEISM MODEL TABLE
# ============================================================

df_abs_model = df_abs_filtered[
    [
        "period_enddate",
        "year",
        "quarter",
        "BedrijfskenmerkenSBI2008_Title",
        "Ziekteverzuimpercentage_1"
    ]
].copy()

df_abs_model.columns = [
    "period_enddate",
    "year",
    "quarter",
    "sector",
    "absenteeism"
]

display(df_abs_model.head())

,period_enddate,year,quarter,sector,absenteeism
0,1996-03-31,1996,1,A-U Alle economische activiteiten,5.5
1,1996-06-30,1996,2,A-U Alle economische activiteiten,4.6
2,1996-09-30,1996,3,A-U Alle economische activiteiten,4.0
3,1996-12-31,1996,4,A-U Alle economische activiteiten,4.7
5,1997-03-31,1997,1,A-U Alle economische activiteiten,4.9


In [74]:
# ============================================================
# LABOR MODEL TABLE
# ============================================================

# ============================================================
# KEEP ONLY TOTAL WORKFORCE
# ============================================================

df_labor_filtered = (
    df_labor_filtered[
        df_labor_filtered[
            "TypeWerkenden_Title"
        ] == "Totaal"
    ]
    .copy()
)

print(df_labor_filtered.shape)

display(
    df_labor_filtered[
        [
            "TypeWerkenden_Title"
        ]
    ].drop_duplicates()
)
df_labor_model = df_labor_filtered[
    [
        "period_enddate",
        "BedrijfstakkenBranchesSBI2008_Title",
        "WerkzamePersonenSeizoengecorrigeerd_2",
        "BanenSeizoengecorrigeerd_4",
        "GewerkteUrenSeizoengecorrigeerd_6"
    ]
].copy()

df_labor_model.columns = [
    "period_enddate",
    "sector",
    "labor_persons",
    "labor_jobs",
    "labor_hours"
]

display(df_labor_model.head())

(1240, 29)


,TypeWerkenden_Title
0,Totaal


,period_enddate,sector,labor_persons,labor_jobs,labor_hours
0,1995-03-31,A-U Alle economische activiteiten,7215.0,8000.0,2651.0
1,1995-06-30,A-U Alle economische activiteiten,7249.0,8043.0,2708.0
2,1995-09-30,A-U Alle economische activiteiten,7297.0,8097.0,2758.0
3,1995-12-31,A-U Alle economische activiteiten,7358.0,8167.0,2826.0
5,1996-03-31,A-U Alle economische activiteiten,7367.0,8187.0,2761.0


In [75]:
# ============================================================
# COMPENSATION MODEL TABLE
# ============================================================

df_comp_model = df_comp_filtered[
    [
        "period_enddate",
        "BedrijfstakkenBranchesSBI2008_Title",
        "BeloningSeizoengecorrigeerd_2",
        "LonenSeizoengecorrigeerd_4",
        "Loonkosten_7"
    ]
].copy()

df_comp_model.columns = [
    "period_enddate",
    "sector",
    "compensation",
    "wages",
    "labor_costs"
]

display(df_comp_model.head())

,period_enddate,sector,compensation,wages,labor_costs
0,1995-03-31,A-U Alle economische activiteiten,40204.0,33627.0,38379.0
1,1995-06-30,A-U Alle economische activiteiten,41191.0,34562.0,46047.0
2,1995-09-30,A-U Alle economische activiteiten,41735.0,35059.0,39202.0
3,1995-12-31,A-U Alle economische activiteiten,42475.0,35743.0,41609.0
5,1996-03-31,A-U Alle economische activiteiten,41980.0,35104.0,39979.0


In [76]:
# ============================================================
# MERGE ALL TABLES
# ============================================================

df_merged = (
    df_abs_model
    .merge(
        df_labor_model,
        on=[
            "period_enddate",
            "sector"
        ],
        how="left"
    )
    .merge(
        df_comp_model,
        on=[
            "period_enddate",
            "sector"
        ],
        how="left"
    )
)

# ------------------------------------------------------------
# Sort
# ------------------------------------------------------------

df_merged = (
    df_merged
    .sort_values(
        [
            "sector",
            "period_enddate"
        ]
    )
    .reset_index(drop=True)
)

print(df_merged.shape)

display(df_merged.head())

# ============================================================
# VALIDATION
# ============================================================

print("\nMissing Values:\n")

print(
    df_merged.isna().sum()
)

print("\nDate Range:")

print(
    df_merged["period_enddate"].min()
)

print(
    df_merged["period_enddate"].max()
)

print("\nNumber of sectors:")

print(
    df_merged["sector"].nunique()
)

(1200, 11)


,period_enddate,year,quarter,sector,absenteeism,labor_persons,labor_jobs,labor_hours,compensation,wages,labor_costs
0,1996-03-31,1996,1,"A Landbouw, bosbouw en visserij",4.5,235.0,295.0,110.0,465.0,398.0,426.0
1,1996-06-30,1996,2,"A Landbouw, bosbouw en visserij",3.8,237.0,297.0,111.0,471.0,402.0,532.0
2,1996-09-30,1996,3,"A Landbouw, bosbouw en visserij",2.7,238.0,297.0,112.0,472.0,405.0,464.0
3,1996-12-31,1996,4,"A Landbouw, bosbouw en visserij",3.2,238.0,297.0,113.0,473.0,405.0,438.0
4,1997-03-31,1997,1,"A Landbouw, bosbouw en visserij",4.0,241.0,299.0,111.0,471.0,407.0,430.0



Missing Values:

period_enddate     0
year               0
quarter            0
sector             0
absenteeism       64
labor_persons      0
labor_jobs         0
labor_hours        0
compensation      10
wages             10
labor_costs       10
dtype: int64

Date Range:
1996-03-31 00:00:00
2025-12-31 00:00:00

Number of sectors:
10


In [77]:
# ============================================================
# MISSING ABSENTEEISM BY SECTOR
# ============================================================

missing_abs = (
    df_merged
    .groupby("sector")["absenteeism"]
    .apply(lambda x: x.isna().sum())
    .reset_index()
)

display(missing_abs)

# ============================================================
# MISSING COMPENSATION BY SECTOR
# ============================================================

missing_comp = (
    df_merged
    .groupby("sector")[
        [
            "compensation",
            "wages",
            "labor_costs"
        ]
    ]
    .apply(lambda x: x.isna().sum())
)

display(missing_comp)

,sector,absenteeism
0,"A Landbouw, bosbouw en visserij",0
1,A-U Alle economische activiteiten,0
2,C Industrie,0
3,F Bouwnijverheid,0
4,J Informatie en communicatie,16
5,K Financiële dienstverlening,0
6,L Verhuur en handel van onroerend goed,16
7,O Openbaar bestuur en overheidsdiensten,0
8,P Onderwijs,32
9,Q Gezondheids- en welzijnszorg,0


,compensation,wages,labor_costs
sector,,,
"A Landbouw, bosbouw en visserij",1,1,1
A-U Alle economische activiteiten,1,1,1
C Industrie,1,1,1
F Bouwnijverheid,1,1,1
J Informatie en communicatie,1,1,1
K Financiële dienstverlening,1,1,1
L Verhuur en handel van onroerend goed,1,1,1
O Openbaar bestuur en overheidsdiensten,1,1,1
P Onderwijs,1,1,1


In [78]:
# ============================================================
# KEEP 2010+
# ============================================================

df_merged = (
    df_merged[
        df_merged["year"] >= 2010
    ]
    .copy()
)

# ============================================================
# DROP MISSING TARGET
# ============================================================

df_merged = (
    df_merged[
        df_merged["absenteeism"].notna()
    ]
    .copy()
)

# ============================================================
# SORT
# ============================================================

df_merged = (
    df_merged
    .sort_values(
        [
            "sector",
            "period_enddate"
        ]
    )
    .reset_index(drop=True)
)

print(df_merged.shape)

print("\nDate Range:")

print(
    df_merged["period_enddate"].min()
)

print(
    df_merged["period_enddate"].max()
)

print("\nMissing Values:\n")

print(
    df_merged.isna().sum()
)

(640, 11)

Date Range:
2010-03-31 00:00:00
2025-12-31 00:00:00

Missing Values:

period_enddate     0
year               0
quarter            0
sector             0
absenteeism        0
labor_persons      0
labor_jobs         0
labor_hours        0
compensation      10
wages             10
labor_costs       10
dtype: int64


In [79]:
# ============================================================
# FORWARD FILL ECONOMIC FEATURES
# ============================================================

economic_cols = [
    "compensation",
    "wages",
    "labor_costs"
]

df_merged[economic_cols] = (
    df_merged
    .groupby("sector")[economic_cols]
    .ffill()
)

# ============================================================
# VALIDATION
# ============================================================

print(
    df_merged[economic_cols]
    .isna()
    .sum()
)

compensation    0
wages           0
labor_costs     0
dtype: int64


In [80]:
# ============================================================
# FEATURE ENGINEERING
# ============================================================

df_features = df_merged.copy()

# ============================================================
# ABSENTEEISM LAGS
# ============================================================

for lag in [1, 2, 3, 4]:

    df_features[f"lag_{lag}"] = (
        df_features
        .groupby("sector")["absenteeism"]
        .shift(lag)
    )

# ============================================================
# ROLLING FEATURES
# ============================================================

df_features["rolling_mean_4"] = (
    df_features
    .groupby("sector")["absenteeism"]
    .transform(
        lambda x:
        x.shift(1).rolling(4).mean()
    )
)

df_features["rolling_std_4"] = (
    df_features
    .groupby("sector")["absenteeism"]
    .transform(
        lambda x:
        x.shift(1).rolling(4).std()
    )
)

# ============================================================
# ECONOMIC LAG FEATURES
# ============================================================

economic_vars = [
    "labor_persons",
    "labor_jobs",
    "labor_hours",
    "compensation",
    "wages",
    "labor_costs"
]

for col in economic_vars:

    df_features[f"{col}_lag_1"] = (
        df_features
        .groupby("sector")[col]
        .shift(1)
    )

# ============================================================
# TIME INDEX
# ============================================================

df_features["time_index"] = np.arange(
    len(df_features)
)

# ============================================================
# TARGET
# NEXT QUARTER ABSENTEEISM
# ============================================================

df_features["target"] = (
    df_features
    .groupby("sector")["absenteeism"]
    .shift(-1)
)

# ============================================================
# DROP MISSING ROWS
# ============================================================

df_features = (
    df_features
    .dropna()
    .reset_index(drop=True)
)

print(df_features.shape)

display(df_features.head())

(590, 25)


,period_enddate,year,quarter,sector,absenteeism,labor_persons,labor_jobs,labor_hours,compensation,wages,...,rolling_mean_4,rolling_std_4,labor_persons_lag_1,labor_jobs_lag_1,labor_hours_lag_1,compensation_lag_1,wages_lag_1,labor_costs_lag_1,time_index,target
0,2011-03-31,2011,1,"A Landbouw, bosbouw en visserij",3.3,199.0,246.0,95.0,715.0,569.0,...,2.575,0.170783,200.0,246.0,97.0,729.0,588.0,689.0,4,2.7
1,2011-06-30,2011,2,"A Landbouw, bosbouw en visserij",2.7,200.0,249.0,96.0,720.0,579.0,...,2.775,0.386221,199.0,246.0,95.0,715.0,569.0,631.0,5,3.1
2,2011-09-30,2011,3,"A Landbouw, bosbouw en visserij",3.1,202.0,251.0,99.0,729.0,582.0,...,2.800,0.374166,200.0,249.0,96.0,720.0,579.0,793.0,6,3.1
3,2011-12-31,2011,4,"A Landbouw, bosbouw en visserij",3.1,200.0,249.0,96.0,731.0,584.0,...,2.975,0.275379,202.0,251.0,99.0,729.0,582.0,725.0,7,2.8
4,2012-03-31,2012,1,"A Landbouw, bosbouw en visserij",2.8,203.0,250.0,98.0,740.0,585.0,...,3.050,0.251661,200.0,249.0,96.0,731.0,584.0,697.0,8,2.8


In [81]:
# ============================================================
# ONE-HOT ENCODE SECTORS
# ============================================================

df_model = pd.get_dummies(
    df_features,
    columns=["sector"],
    drop_first=True
)

print(df_model.shape)

display(df_model.head())

(590, 33)


,period_enddate,year,quarter,absenteeism,labor_persons,labor_jobs,labor_hours,compensation,wages,labor_costs,...,target,sector_A-U Alle economische activiteiten,sector_C Industrie,sector_F Bouwnijverheid,sector_J Informatie en communicatie,sector_K Financiële dienstverlening,sector_L Verhuur en handel van onroerend goed,sector_O Openbaar bestuur en overheidsdiensten,sector_P Onderwijs,sector_Q Gezondheids- en welzijnszorg
0,2011-03-31,2011,1,3.3,199.0,246.0,95.0,715.0,569.0,631.0,...,2.7,False,False,False,False,False,False,False,False,False
1,2011-06-30,2011,2,2.7,200.0,249.0,96.0,720.0,579.0,793.0,...,3.1,False,False,False,False,False,False,False,False,False
2,2011-09-30,2011,3,3.1,202.0,251.0,99.0,729.0,582.0,725.0,...,3.1,False,False,False,False,False,False,False,False,False
3,2011-12-31,2011,4,3.1,200.0,249.0,96.0,731.0,584.0,697.0,...,2.8,False,False,False,False,False,False,False,False,False
4,2012-03-31,2012,1,2.8,203.0,250.0,98.0,740.0,585.0,655.0,...,2.8,False,False,False,False,False,False,False,False,False


In [82]:
# ============================================================
# BASELINE FEATURES
# ============================================================

baseline_features = [
    "quarter",
    "time_index",
    "lag_1",
    "lag_2",
    "lag_3",
    "lag_4",
    "rolling_mean_4",
    "rolling_std_4"
]

# ============================================================
# EXTENDED FEATURES
# ============================================================

economic_features = [
    "labor_persons_lag_1",
    "labor_jobs_lag_1",
    "labor_hours_lag_1",
    "compensation_lag_1",
    "wages_lag_1",
    "labor_costs_lag_1"
]

# ------------------------------------------------------------
# Sector dummy columns
# ------------------------------------------------------------

sector_dummies = [
    col
    for col in df_model.columns
    if col.startswith("sector_")
]

# ------------------------------------------------------------
# Final feature list
# ------------------------------------------------------------

extended_features = (
    baseline_features
    + economic_features
    + sector_dummies
)

print(len(extended_features))

# ============================================================
# TARGET
# ============================================================

TARGET = "target"

23


In [83]:
# ============================================================
# IMPORTS
# ============================================================

from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import TimeSeriesSplit
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error
)

# ============================================================
# BASELINE FEATURES
# ============================================================

baseline_features = [
    "quarter",
    "time_index",
    "lag_1",
    "lag_2",
    "lag_3",
    "lag_4",
    "rolling_mean_4",
    "rolling_std_4"
]

# ============================================================
# TARGET
# ============================================================

TARGET = "target"

# ============================================================
# TIME SERIES CV
# ============================================================

tscv = TimeSeriesSplit(
    n_splits=5
)

# ============================================================
# STORAGE
# ============================================================

results_baseline = []

# ============================================================
# FORCE NUMERIC TARGET
# ============================================================

df_model["target"] = pd.to_numeric(
    df_model["target"],
    errors="coerce"
)

df_model["absenteeism"] = pd.to_numeric(
    df_model["absenteeism"],
    errors="coerce"
)

print(df_model[["target", "absenteeism"]].dtypes)

# ============================================================
# CV LOOP
# ============================================================

for fold, (train_idx, test_idx) in enumerate(
    tscv.split(df_model),
    start=1
):

    print(f"\nFold {fold}")

    # --------------------------------------------------------
    # SPLIT
    # --------------------------------------------------------

    train_fold = df_model.iloc[train_idx]

    test_fold = df_model.iloc[test_idx]

    # --------------------------------------------------------
    # FEATURES
    # --------------------------------------------------------

    X_train = train_fold[baseline_features]

    y_train = train_fold[TARGET]

    X_test = test_fold[baseline_features]

    y_test = test_fold[TARGET]

    # --------------------------------------------------------
    # MODEL
    # --------------------------------------------------------

    model_rf = RandomForestRegressor(
        n_estimators=300,
        max_depth=8,
        min_samples_leaf=5,
        random_state=42,
        n_jobs=-1
    )

    # --------------------------------------------------------
    # TRAIN
    # --------------------------------------------------------

    model_rf.fit(
        X_train,
        y_train
    )

    # --------------------------------------------------------
    # PREDICT
    # --------------------------------------------------------

    pred = model_rf.predict(X_test)

    # --------------------------------------------------------
    # METRICS
    # --------------------------------------------------------

    mae = mean_absolute_error(
        y_test,
        pred
    )

    rmse = np.sqrt(
        mean_squared_error(
            y_test,
            pred
        )
    )

    mape = np.mean(
        np.abs(
            (y_test - pred)
            / y_test
        )
    ) * 100

    # --------------------------------------------------------
    # STORE
    # --------------------------------------------------------

    results_baseline.append({
        "fold": fold,
        "MAE": mae,
        "RMSE": rmse,
        "MAPE": mape
    })

# ============================================================
# RESULTS
# ============================================================

results_baseline = pd.DataFrame(
    results_baseline
).round(3)

display(results_baseline)

# ============================================================
# AVERAGE
# ============================================================

summary_baseline = (
    results_baseline[
        [
            "MAE",
            "RMSE",
            "MAPE"
        ]
    ]
    .mean()
    .to_frame()
    .T
)

summary_baseline.insert(
    0,
    "Model",
    "RF Baseline"
)

summary_baseline = summary_baseline.round(3)

print("\nAVERAGE PERFORMANCE")

display(summary_baseline)

target         float64
absenteeism    float64
dtype: object

Fold 1

Fold 2

Fold 3

Fold 4

Fold 5


,fold,MAE,RMSE,MAPE
0,1,0.720,0.930,12.915
1,2,0.425,0.521,12.681
2,3,0.250,0.296,8.384
3,4,0.313,0.411,6.391
4,5,0.392,0.604,6.141



AVERAGE PERFORMANCE


,Model,MAE,RMSE,MAPE
0,RF Baseline,0.42,0.552,9.302


In [84]:
# ============================================================
# EXTENDED FEATURES
# ============================================================

economic_features = [
    "labor_persons_lag_1",
    "labor_jobs_lag_1",
    "labor_hours_lag_1",
    "compensation_lag_1",
    "wages_lag_1",
    "labor_costs_lag_1"
]

sector_dummies = [
    col
    for col in df_model.columns
    if col.startswith("sector_")
]

extended_features = (
    baseline_features
    + economic_features
    + sector_dummies
)

print(len(extended_features))

# ============================================================
# STORAGE
# ============================================================

results_extended = []

# ============================================================
# CV LOOP
# ============================================================

for fold, (train_idx, test_idx) in enumerate(
    tscv.split(df_model),
    start=1
):

    print(f"\nFold {fold}")

    # --------------------------------------------------------
    # SPLIT
    # --------------------------------------------------------

    train_fold = df_model.iloc[train_idx]

    test_fold = df_model.iloc[test_idx]

    # --------------------------------------------------------
    # FEATURES
    # --------------------------------------------------------

    X_train = train_fold[extended_features]

    y_train = train_fold[TARGET]

    X_test = test_fold[extended_features]

    y_test = test_fold[TARGET]

    # --------------------------------------------------------
    # MODEL
    # --------------------------------------------------------

    model_rf_ext = RandomForestRegressor(
        n_estimators=300,
        max_depth=8,
        min_samples_leaf=5,
        random_state=42,
        n_jobs=-1
    )

    # --------------------------------------------------------
    # TRAIN
    # --------------------------------------------------------

    model_rf_ext.fit(
        X_train,
        y_train
    )

    # --------------------------------------------------------
    # PREDICT
    # --------------------------------------------------------

    pred = model_rf_ext.predict(X_test)

    # --------------------------------------------------------
    # METRICS
    # --------------------------------------------------------

    mae = mean_absolute_error(
        y_test,
        pred
    )

    rmse = np.sqrt(
        mean_squared_error(
            y_test,
            pred
        )
    )

    mape = np.mean(
        np.abs(
            (y_test - pred)
            / y_test
        )
    ) * 100

    # --------------------------------------------------------
    # STORE
    # --------------------------------------------------------

    results_extended.append({
        "fold": fold,
        "MAE": mae,
        "RMSE": rmse,
        "MAPE": mape
    })

# ============================================================
# RESULTS
# ============================================================

results_extended = pd.DataFrame(
    results_extended
).round(3)

display(results_extended)

# ============================================================
# AVERAGE
# ============================================================

summary_extended = (
    results_extended[
        [
            "MAE",
            "RMSE",
            "MAPE"
        ]
    ]
    .mean()
    .to_frame()
    .T
)

summary_extended.insert(
    0,
    "Model",
    "RF Extended"
)

summary_extended = summary_extended.round(3)

print("\nAVERAGE PERFORMANCE")

display(summary_extended)

23

Fold 1

Fold 2

Fold 3

Fold 4

Fold 5


,fold,MAE,RMSE,MAPE
0,1,0.850,1.100,15.014
1,2,0.439,0.547,13.392
2,3,0.252,0.309,8.074
3,4,0.319,0.420,6.476
4,5,0.415,0.619,6.615



AVERAGE PERFORMANCE


,Model,MAE,RMSE,MAPE
0,RF Extended,0.455,0.599,9.914


In [85]:
# ============================================================
# FORCE ECONOMIC COLUMNS TO NUMERIC
# ============================================================

economic_vars = [
    "labor_persons",
    "labor_jobs",
    "labor_hours",
    "compensation",
    "wages",
    "labor_costs"
]

for col in economic_vars:

    df_merged[col] = pd.to_numeric(
        df_merged[col],
        errors="coerce"
    )

# ============================================================
# VALIDATION
# ============================================================

print(
    df_merged[economic_vars]
    .dtypes
)

labor_persons    float64
labor_jobs       float64
labor_hours      float64
compensation     float64
wages            float64
labor_costs      float64
dtype: object


In [86]:
# ============================================================
# COPY
# ============================================================

df_growth = df_merged.copy()

# ============================================================
# VARIABLES
# ============================================================

economic_vars = [
    "labor_persons",
    "labor_jobs",
    "labor_hours",
    "compensation",
    "wages",
    "labor_costs"
]

# ============================================================
# QUARTER-OVER-QUARTER GROWTH
# ============================================================

for col in economic_vars:

    df_growth[f"{col}_qoq_growth"] = (
        df_growth
        .groupby("sector")[col]
        .pct_change(1)
    )

# ============================================================
# YEAR-OVER-YEAR GROWTH
# ============================================================

for col in economic_vars:

    df_growth[f"{col}_yoy_growth"] = (
        df_growth
        .groupby("sector")[col]
        .pct_change(4)
    )

# ============================================================
# ABSENTEEISM FEATURES
# ============================================================

for lag in [1, 2, 3, 4]:

    df_growth[f"lag_{lag}"] = (
        df_growth
        .groupby("sector")["absenteeism"]
        .shift(lag)
    )

# ============================================================
# ROLLING FEATURES
# ============================================================

df_growth["rolling_mean_4"] = (
    df_growth
    .groupby("sector")["absenteeism"]
    .transform(
        lambda x:
        x.shift(1).rolling(4).mean()
    )
)

df_growth["rolling_std_4"] = (
    df_growth
    .groupby("sector")["absenteeism"]
    .transform(
        lambda x:
        x.shift(1).rolling(4).std()
    )
)

# ============================================================
# TIME INDEX
# ============================================================

df_growth["time_index"] = np.arange(
    len(df_growth)
)

# ============================================================
# TARGET
# ============================================================

df_growth["target"] = (
    df_growth
    .groupby("sector")["absenteeism"]
    .shift(-1)
)

# ============================================================
# DROP NA
# ============================================================

df_growth = (
    df_growth
    .dropna()
    .reset_index(drop=True)
)

print(df_growth.shape)

display(df_growth.head())

(590, 31)


,period_enddate,year,quarter,sector,absenteeism,labor_persons,labor_jobs,labor_hours,compensation,wages,...,wages_yoy_growth,labor_costs_yoy_growth,lag_1,lag_2,lag_3,lag_4,rolling_mean_4,rolling_std_4,time_index,target
0,2011-03-31,2011,1,"A Landbouw, bosbouw en visserij",3.3,199.0,246.0,95.0,715.0,569.0,...,0.017889,0.039539,2.8,2.4,2.6,2.5,2.575,0.170783,4,2.7
1,2011-06-30,2011,2,"A Landbouw, bosbouw en visserij",2.7,200.0,249.0,96.0,720.0,579.0,...,0.026596,0.033898,3.3,2.8,2.4,2.6,2.775,0.386221,5,3.1
2,2011-09-30,2011,3,"A Landbouw, bosbouw en visserij",3.1,202.0,251.0,99.0,729.0,582.0,...,0.015707,0.035714,2.7,3.3,2.8,2.4,2.800,0.374166,6,3.1
3,2011-12-31,2011,4,"A Landbouw, bosbouw en visserij",3.1,200.0,249.0,96.0,731.0,584.0,...,-0.006803,0.011611,3.1,2.7,3.3,2.8,2.975,0.275379,7,2.8
4,2012-03-31,2012,1,"A Landbouw, bosbouw en visserij",2.8,203.0,250.0,98.0,740.0,585.0,...,0.028120,0.038035,3.1,3.1,2.7,3.3,3.050,0.251661,8,2.8


In [87]:
# ============================================================
# ONE-HOT ENCODE SECTORS
# ============================================================

df_growth_model = pd.get_dummies(
    df_growth,
    columns=["sector"],
    drop_first=True
)

print(df_growth_model.shape)

# ============================================================
# BASE FEATURES
# ============================================================

base_features = [
    "quarter",
    "time_index",
    "lag_1",
    "lag_2",
    "lag_3",
    "lag_4",
    "rolling_mean_4",
    "rolling_std_4"
]

# ============================================================
# GROWTH FEATURES
# ============================================================

growth_features = [
    col
    for col in df_growth_model.columns
    if (
        "qoq_growth" in col
        or
        "yoy_growth" in col
    )
]

# ============================================================
# SECTOR DUMMIES
# ============================================================

sector_features = [
    col
    for col in df_growth_model.columns
    if col.startswith("sector_")
]

# ============================================================
# FINAL FEATURES
# ============================================================

rf_growth_features = (
    base_features
    + growth_features
    + sector_features
)

print(len(rf_growth_features))

# ============================================================
# FORCE TARGET TO NUMERIC
# ============================================================

df_growth_model["target"] = pd.to_numeric(
    df_growth_model["target"],
    errors="coerce"
)

df_growth_model["absenteeism"] = pd.to_numeric(
    df_growth_model["absenteeism"],
    errors="coerce"
)

print(
    df_growth_model[
        ["target", "absenteeism"]
    ].dtypes
)

# ============================================================
# TARGET
# ============================================================

TARGET = "target"

# ============================================================
# STORAGE
# ============================================================

results_growth_rf = []

# ============================================================
# TIME SERIES CV
# ============================================================

tscv = TimeSeriesSplit(
    n_splits=5
)

# ============================================================
# CV LOOP
# ============================================================

for fold, (train_idx, test_idx) in enumerate(
    tscv.split(df_growth_model),
    start=1
):

    print(f"\nFold {fold}")

    # --------------------------------------------------------
    # SPLIT
    # --------------------------------------------------------

    train_fold = df_growth_model.iloc[train_idx]

    test_fold = df_growth_model.iloc[test_idx]

    # --------------------------------------------------------
    # FEATURES
    # --------------------------------------------------------

    X_train = train_fold[rf_growth_features]

    y_train = train_fold[TARGET]

    X_test = test_fold[rf_growth_features]

    y_test = test_fold[TARGET]

    # --------------------------------------------------------
    # MODEL
    # --------------------------------------------------------

    model_growth_rf = RandomForestRegressor(
        n_estimators=300,
        max_depth=8,
        min_samples_leaf=5,
        random_state=42,
        n_jobs=-1
    )

    # --------------------------------------------------------
    # TRAIN
    # --------------------------------------------------------

    model_growth_rf.fit(
        X_train,
        y_train
    )

    # --------------------------------------------------------
    # PREDICT
    # --------------------------------------------------------

    pred = model_growth_rf.predict(X_test)

    # --------------------------------------------------------
    # METRICS
    # --------------------------------------------------------

    mae = mean_absolute_error(
        y_test,
        pred
    )

    rmse = np.sqrt(
        mean_squared_error(
            y_test,
            pred
        )
    )

    mape = np.mean(
        np.abs(
            (y_test - pred)
            / y_test
        )
    ) * 100

    # --------------------------------------------------------
    # STORE
    # --------------------------------------------------------

    results_growth_rf.append({
        "fold": fold,
        "MAE": mae,
        "RMSE": rmse,
        "MAPE": mape
    })

# ============================================================
# RESULTS
# ============================================================

results_growth_rf = pd.DataFrame(
    results_growth_rf
).round(3)

display(results_growth_rf)

# ============================================================
# SUMMARY
# ============================================================

summary_growth_rf = (
    results_growth_rf[
        [
            "MAE",
            "RMSE",
            "MAPE"
        ]
    ]
    .mean()
    .to_frame()
    .T
)

summary_growth_rf.insert(
    0,
    "Model",
    "RF Growth Features"
)

summary_growth_rf = summary_growth_rf.round(3)

print("\nAVERAGE PERFORMANCE")

display(summary_growth_rf)

(590, 39)
29
target         float64
absenteeism    float64
dtype: object

Fold 1

Fold 2

Fold 3

Fold 4

Fold 5


,fold,MAE,RMSE,MAPE
0,1,0.737,0.957,13.134
1,2,0.427,0.524,12.706
2,3,0.245,0.291,8.103
3,4,0.302,0.411,6.089
4,5,0.397,0.610,6.220



AVERAGE PERFORMANCE


,Model,MAE,RMSE,MAPE
0,RF Growth Features,0.422,0.559,9.25


In [88]:
# ============================================================
# STORAGE
# ============================================================

results_gb = []

# ============================================================
# TIME SERIES CV
# ============================================================

tscv = TimeSeriesSplit(
    n_splits=5
)

# ============================================================
# CV LOOP
# ============================================================

for fold, (train_idx, test_idx) in enumerate(
    tscv.split(df_growth_model),
    start=1
):

    print(f"\nFold {fold}")

    # --------------------------------------------------------
    # SPLIT
    # --------------------------------------------------------

    train_fold = df_growth_model.iloc[train_idx]

    test_fold = df_growth_model.iloc[test_idx]

    # --------------------------------------------------------
    # FEATURES
    # --------------------------------------------------------

    X_train = train_fold[rf_growth_features]

    y_train = train_fold[TARGET]

    X_test = test_fold[rf_growth_features]

    y_test = test_fold[TARGET]

    # --------------------------------------------------------
    # MODEL
    # --------------------------------------------------------

    model_gb = GradientBoostingRegressor(
        n_estimators=300,
        learning_rate=0.03,
        max_depth=4,
        random_state=42
    )

    # --------------------------------------------------------
    # TRAIN
    # --------------------------------------------------------

    model_gb.fit(
        X_train,
        y_train
    )

    # --------------------------------------------------------
    # PREDICT
    # --------------------------------------------------------

    pred = model_gb.predict(X_test)

    # --------------------------------------------------------
    # METRICS
    # --------------------------------------------------------

    mae = mean_absolute_error(
        y_test,
        pred
    )

    rmse = np.sqrt(
        mean_squared_error(
            y_test,
            pred
        )
    )

    mape = np.mean(
        np.abs(
            (y_test - pred)
            / y_test
        )
    ) * 100

    # --------------------------------------------------------
    # STORE
    # --------------------------------------------------------

    results_gb.append({
        "fold": fold,
        "MAE": mae,
        "RMSE": rmse,
        "MAPE": mape
    })

# ============================================================
# RESULTS
# ============================================================

results_gb = pd.DataFrame(
    results_gb
).round(3)

display(results_gb)

# ============================================================
# SUMMARY
# ============================================================

summary_gb = (
    results_gb[
        [
            "MAE",
            "RMSE",
            "MAPE"
        ]
    ]
    .mean()
    .to_frame()
    .T
)

summary_gb.insert(
    0,
    "Model",
    "Gradient Boosting Growth"
)

summary_gb = summary_gb.round(3)

print("\nAVERAGE PERFORMANCE")

display(summary_gb)




Fold 1

Fold 2

Fold 3

Fold 4

Fold 5


,fold,MAE,RMSE,MAPE
0,1,0.639,0.840,11.505
1,2,0.428,0.529,12.300
2,3,0.246,0.294,8.201
3,4,0.378,0.474,7.281
4,5,0.406,0.593,6.436



AVERAGE PERFORMANCE


,Model,MAE,RMSE,MAPE
0,Gradient Boosting Growth,0.419,0.546,9.145


In [89]:
# ============================================================
# TRAIN FINAL GB MODEL ON FULL DATA
# ============================================================

X_full = df_growth_model[rf_growth_features]

y_full = df_growth_model[TARGET]

# ============================================================
# MODEL
# ============================================================

final_gb = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.03,
    max_depth=4,
    random_state=42
)

# ============================================================
# TRAIN
# ============================================================

final_gb.fit(
    X_full,
    y_full
)

# ============================================================
# FEATURE IMPORTANCE
# ============================================================

importance_df = pd.DataFrame({
    "feature": rf_growth_features,
    "importance": final_gb.feature_importances_
})

# ============================================================
# SORT
# ============================================================

importance_df = (
    importance_df
    .sort_values(
        "importance",
        ascending=False
    )
    .reset_index(drop=True)
)

# ============================================================
# PERCENTAGE
# ============================================================

importance_df["importance_pct"] = (
    importance_df["importance"]
    * 100
)

importance_df = importance_df.round(3)

# ============================================================
# DISPLAY
# ============================================================

display(
    importance_df.head(20)
)

,feature,importance,importance_pct
0,rolling_mean_4,0.456,45.604
1,lag_3,0.432,43.182
2,time_index,0.033,3.297
3,quarter,0.021,2.126
4,lag_2,0.014,1.401
5,rolling_std_4,0.010,0.996
6,labor_costs_qoq_growth,0.006,0.577
7,lag_1,0.004,0.353
8,labor_persons_yoy_growth,0.003,0.317
9,wages_yoy_growth,0.003,0.299


In [90]:
# ============================================================
# FEATURES
# ============================================================

local_features = [
    "quarter",
    "time_index",
    "lag_1",
    "lag_2",
    "lag_3",
    "lag_4",
    "rolling_mean_4",
    "rolling_std_4"
]

growth_features = [
    col
    for col in df_growth.columns
    if (
        "qoq_growth" in col
        or
        "yoy_growth" in col
    )
]

local_features = (
    local_features
    + growth_features
)

# ============================================================
# STORAGE
# ============================================================

all_sector_importance = []

# ============================================================
# LOOP THROUGH SECTORS
# ============================================================

for sector_name in sorted(df_growth["sector"].unique()):

    print(f"\nProcessing: {sector_name}")

    # --------------------------------------------------------
    # FILTER
    # --------------------------------------------------------

    df_sector = (
        df_growth[
            df_growth["sector"] == sector_name
        ]
        .copy()
    )

    # --------------------------------------------------------
    # FEATURES / TARGET
    # --------------------------------------------------------

    X = df_sector[local_features]

    y = df_sector["target"]

    # --------------------------------------------------------
    # MODEL
    # --------------------------------------------------------

    model_local = GradientBoostingRegressor(
        n_estimators=300,
        learning_rate=0.03,
        max_depth=4,
        random_state=42
    )

    # --------------------------------------------------------
    # TRAIN
    # --------------------------------------------------------

    model_local.fit(
        X,
        y
    )

    # --------------------------------------------------------
    # IMPORTANCE
    # --------------------------------------------------------

    importance_df = pd.DataFrame({
        "feature": local_features,
        "importance": model_local.feature_importances_
    })

    importance_df["sector"] = sector_name

    # --------------------------------------------------------
    # SORT
    # --------------------------------------------------------

    importance_df = (
        importance_df
        .sort_values(
            "importance",
            ascending=False
        )
        .reset_index(drop=True)
    )

    # --------------------------------------------------------
    # PERCENT
    # --------------------------------------------------------

    importance_df["importance_pct"] = (
        importance_df["importance"]
        * 100
    )

    # --------------------------------------------------------
    # KEEP TOP 10
    # --------------------------------------------------------

    importance_df = (
        importance_df
        .head(10)
    )

    all_sector_importance.append(
        importance_df
    )

# ============================================================
# COMBINE
# ============================================================

all_sector_importance = pd.concat(
    all_sector_importance,
    ignore_index=True
)

all_sector_importance = (
    all_sector_importance
    .round(3)
)

# ============================================================
# DISPLAY
# ============================================================

display(all_sector_importance)


Processing: A Landbouw, bosbouw en visserij

Processing: A-U Alle economische activiteiten

Processing: C Industrie

Processing: F Bouwnijverheid

Processing: J Informatie en communicatie

Processing: K Financiële dienstverlening

Processing: L Verhuur en handel van onroerend goed

Processing: O Openbaar bestuur en overheidsdiensten

Processing: P Onderwijs

Processing: Q Gezondheids- en welzijnszorg


,feature,importance,sector,importance_pct
0,time_index,0.787,"A Landbouw, bosbouw en visserij",78.718
1,compensation_yoy_growth,0.048,"A Landbouw, bosbouw en visserij",4.828
2,wages_yoy_growth,0.036,"A Landbouw, bosbouw en visserij",3.569
3,rolling_std_4,0.027,"A Landbouw, bosbouw en visserij",2.690
4,labor_hours_yoy_growth,0.025,"A Landbouw, bosbouw en visserij",2.451
...,...,...,...,...
95,rolling_std_4,0.010,Q Gezondheids- en welzijnszorg,1.005
96,labor_costs_qoq_growth,0.008,Q Gezondheids- en welzijnszorg,0.824
97,labor_persons_qoq_growth,0.006,Q Gezondheids- en welzijnszorg,0.563
98,wages_yoy_growth,0.005,Q Gezondheids- en welzijnszorg,0.532


In [93]:
# ============================================================
# STORAGE
# ============================================================

sector_results = []

# ============================================================
# LOOP THROUGH SECTORS
# ============================================================

for sector_name in sorted(df_growth["sector"].unique()):

    print(f"\nProcessing: {sector_name}")

    # --------------------------------------------------------
    # FILTER
    # --------------------------------------------------------

    df_sector = (
        df_growth[
            df_growth["sector"] == sector_name
        ]
        .copy()
        .reset_index(drop=True)
    )

    # --------------------------------------------------------
    # FORCE NUMERIC
    # --------------------------------------------------------

    df_sector["target"] = pd.to_numeric(
        df_sector["target"],
        errors="coerce"
    )

    df_sector["absenteeism"] = pd.to_numeric(
        df_sector["absenteeism"],
        errors="coerce"
    )

    # --------------------------------------------------------
    # DROP NA TARGETS
    # --------------------------------------------------------

    df_sector = df_sector.dropna(
        subset=["target"]
    )

    # --------------------------------------------------------
    # STORAGE FOR CV RESULTS
    # --------------------------------------------------------

    fold_mae = []

    fold_rmse = []

    fold_mape = []

    # --------------------------------------------------------
    # TIME SERIES CV
    # --------------------------------------------------------

    tscv = TimeSeriesSplit(
        n_splits=5
    )

    # --------------------------------------------------------
    # CV LOOP
    # --------------------------------------------------------

    for train_idx, test_idx in tscv.split(df_sector):

        # ----------------------------------------------------
        # SPLIT
        # ----------------------------------------------------

        train_fold = df_sector.iloc[train_idx]

        test_fold = df_sector.iloc[test_idx]

        # ----------------------------------------------------
        # FEATURES / TARGET
        # ----------------------------------------------------

        X_train = train_fold[local_features]

        y_train = train_fold["target"]

        X_test = test_fold[local_features]

        y_test = test_fold["target"]

        # ----------------------------------------------------
        # MODEL
        # ----------------------------------------------------

        model_local = GradientBoostingRegressor(
            n_estimators=300,
            learning_rate=0.03,
            max_depth=4,
            random_state=42
        )

        # ----------------------------------------------------
        # TRAIN
        # ----------------------------------------------------

        model_local.fit(
            X_train,
            y_train
        )

        # ----------------------------------------------------
        # PREDICT
        # ----------------------------------------------------

        pred = model_local.predict(X_test)

        # ----------------------------------------------------
        # METRICS
        # ----------------------------------------------------

        mae = mean_absolute_error(
            y_test,
            pred
        )

        rmse = np.sqrt(
            mean_squared_error(
                y_test,
                pred
            )
        )

        mape = np.mean(
            np.abs(
                (y_test - pred)
                / y_test
            )
        ) * 100

        # ----------------------------------------------------
        # STORE FOLD RESULTS
        # ----------------------------------------------------

        fold_mae.append(mae)

        fold_rmse.append(rmse)

        fold_mape.append(mape)

    # --------------------------------------------------------
    # STORE SECTOR RESULTS
    # --------------------------------------------------------

    sector_results.append({
        "sector": sector_name,
        "MAE": np.mean(fold_mae),
        "RMSE": np.mean(fold_rmse),
        "MAPE": np.mean(fold_mape)
    })

# ============================================================
# FINAL RESULTS
# ============================================================

sector_results = pd.DataFrame(
    sector_results
)

sector_results = (
    sector_results
    .sort_values(
        "MAPE"
    )
    .reset_index(drop=True)
)

sector_results = sector_results.round(3)

display(sector_results)


Processing: A Landbouw, bosbouw en visserij

Processing: A-U Alle economische activiteiten

Processing: C Industrie

Processing: F Bouwnijverheid

Processing: J Informatie en communicatie

Processing: K Financiële dienstverlening

Processing: L Verhuur en handel van onroerend goed

Processing: O Openbaar bestuur en overheidsdiensten

Processing: P Onderwijs

Processing: Q Gezondheids- en welzijnszorg


,sector,MAE,RMSE,MAPE
0,P Onderwijs,0.324,0.414,6.006
1,C Industrie,0.382,0.479,6.614
2,A-U Alle economische activiteiten,0.349,0.414,7.049
3,F Bouwnijverheid,0.315,0.376,7.083
4,O Openbaar bestuur en overheidsdiensten,0.436,0.538,7.799
5,Q Gezondheids- en welzijnszorg,0.590,0.696,8.782
6,K Financiële dienstverlening,0.326,0.380,10.048
7,"A Landbouw, bosbouw en visserij",0.329,0.394,10.205
8,L Verhuur en handel van onroerend goed,0.388,0.434,10.300
9,J Informatie en communicatie,0.347,0.405,10.365


In [94]:
# ============================================================
# IMPORTS
# ============================================================

from sklearn.model_selection import GridSearchCV

# ============================================================
# SELECT SECTOR
# ============================================================

sector_name = "P Onderwijs"

# ============================================================
# FILTER DATA
# ============================================================

df_sector = (
    df_growth[
        df_growth["sector"] == sector_name
    ]
    .copy()
    .reset_index(drop=True)
)

# ============================================================
# FORCE NUMERIC
# ============================================================

df_sector["target"] = pd.to_numeric(
    df_sector["target"],
    errors="coerce"
)

df_sector = df_sector.dropna(
    subset=["target"]
)

# ============================================================
# FEATURES / TARGET
# ============================================================

X = df_sector[local_features]

y = df_sector["target"]

# ============================================================
# TIME SERIES CV
# ============================================================

tscv = TimeSeriesSplit(
    n_splits=5
)

# ============================================================
# PARAMETER GRID
# ============================================================

param_grid = {
    "n_estimators": [100, 300],
    "learning_rate": [0.03, 0.05],
    "max_depth": [2, 3, 4],
    "min_samples_leaf": [2, 5]
}

# ============================================================
# BASE MODEL
# ============================================================

gb_model = GradientBoostingRegressor(
    random_state=42
)

# ============================================================
# GRID SEARCH
# ============================================================

grid_search = GridSearchCV(
    estimator=gb_model,
    param_grid=param_grid,
    cv=tscv,
    scoring="neg_mean_absolute_error",
    n_jobs=-1,
    verbose=1
)

# ============================================================
# FIT
# ============================================================

grid_search.fit(
    X,
    y
)

# ============================================================
# BEST RESULTS
# ============================================================

print("\nBEST PARAMETERS:")
print(grid_search.best_params_)

print("\nBEST SCORE:")
print(-grid_search.best_score_)

Fitting 5 folds for each of 24 candidates, totalling 120 fits

BEST PARAMETERS:
{'learning_rate': 0.05, 'max_depth': 2, 'min_samples_leaf': 2, 'n_estimators': 300}

BEST SCORE:
0.2916146441160753


In [95]:
# ============================================================
# SECTORS TO COMPARE
# ============================================================

selected_sectors = [
    "P Onderwijs",
    "J Informatie en communicatie",
    "Q Gezondheids- en welzijnszorg",
    "A Landbouw, bosbouw en visserij"
]

# ============================================================
# STORAGE
# ============================================================

sector_tuning_results = []

# ============================================================
# LOOP THROUGH SECTORS
# ============================================================

for sector_name in selected_sectors:

    print(f"\n{'='*60}")
    print(f"Processing: {sector_name}")
    print(f"{'='*60}")

    # --------------------------------------------------------
    # FILTER
    # --------------------------------------------------------

    df_sector = (
        df_growth[
            df_growth["sector"] == sector_name
        ]
        .copy()
        .reset_index(drop=True)
    )

    # --------------------------------------------------------
    # FORCE NUMERIC
    # --------------------------------------------------------

    df_sector["target"] = pd.to_numeric(
        df_sector["target"],
        errors="coerce"
    )

    df_sector = df_sector.dropna(
        subset=["target"]
    )

    # --------------------------------------------------------
    # FEATURES / TARGET
    # --------------------------------------------------------

    X = df_sector[local_features]

    y = df_sector["target"]

    # --------------------------------------------------------
    # TIME SERIES CV
    # --------------------------------------------------------

    tscv = TimeSeriesSplit(
        n_splits=5
    )

    # --------------------------------------------------------
    # PARAMETER GRID
    # --------------------------------------------------------

    param_grid = {
        "n_estimators": [100, 300],
        "learning_rate": [0.03, 0.05],
        "max_depth": [2, 3, 4],
        "min_samples_leaf": [2, 5]
    }

    # --------------------------------------------------------
    # MODEL
    # --------------------------------------------------------

    gb_model = GradientBoostingRegressor(
        random_state=42
    )

    # --------------------------------------------------------
    # GRID SEARCH
    # --------------------------------------------------------

    grid_search = GridSearchCV(
        estimator=gb_model,
        param_grid=param_grid,
        cv=tscv,
        scoring="neg_mean_absolute_error",
        n_jobs=-1,
        verbose=0
    )

    # --------------------------------------------------------
    # FIT
    # --------------------------------------------------------

    grid_search.fit(
        X,
        y
    )

    # --------------------------------------------------------
    # STORE RESULTS
    # --------------------------------------------------------

    best_params = grid_search.best_params_

    sector_tuning_results.append({
        "sector": sector_name,
        "best_MAE": -grid_search.best_score_,
        "learning_rate": best_params["learning_rate"],
        "max_depth": best_params["max_depth"],
        "min_samples_leaf": best_params["min_samples_leaf"],
        "n_estimators": best_params["n_estimators"]
    })

# ============================================================
# RESULTS TABLE
# ============================================================

sector_tuning_results = pd.DataFrame(
    sector_tuning_results
)

sector_tuning_results = (
    sector_tuning_results
    .sort_values(
        "best_MAE"
    )
    .reset_index(drop=True)
)

sector_tuning_results = sector_tuning_results.round(3)

display(sector_tuning_results)


Processing: P Onderwijs

Processing: J Informatie en communicatie

Processing: Q Gezondheids- en welzijnszorg

Processing: A Landbouw, bosbouw en visserij


,sector,best_MAE,learning_rate,max_depth,min_samples_leaf,n_estimators
0,J Informatie en communicatie,0.279,0.05,4,5,300
1,P Onderwijs,0.292,0.05,2,2,300
2,"A Landbouw, bosbouw en visserij",0.299,0.05,2,2,100
3,Q Gezondheids- en welzijnszorg,0.534,0.05,2,5,100


# Multivariate ML Forecasting with External Economic Features

## Objective

In this notebook, we extended the absenteeism forecasting framework by integrating additional macroeconomic and labor-market related CBS datasets into the machine learning pipeline.

The goal was to evaluate whether external economic indicators improve forecasting accuracy compared to models based only on historical absenteeism dynamics.

---

# Additional CBS Tables Used

| Table ID | Description |
|---|---|
| 85920NED | Labor volume by sector and quarter |
| 85917NED | Compensation and labor costs by sector and quarter |

These tables provided additional explanatory variables related to:
- workforce size,
- number of jobs,
- worked hours,
- wages,
- labor costs,
- compensation dynamics.

---

# Data Integration Workflow

## 1. Loaded Multiple CBS Tables

Three datasets were loaded from the silver database:

- absenteeism table (`80072ned_silver`)
- labor volume table (`85920ned_silver`)
- compensation table (`85917ned_silver`)

---

## 2. Sector Alignment

Only sectors available across all datasets were retained to ensure consistent joins and comparable forecasting behavior.

Final common sectors:

- A Landbouw, bosbouw en visserij
- A-U Alle economische activiteiten
- C Industrie
- F Bouwnijverheid
- J Informatie en communicatie
- K Financiële dienstverlening
- L Verhuur en handel van onroerend goed
- O Openbaar bestuur en overheidsdiensten
- P Onderwijs
- Q Gezondheids- en welzijnszorg

---

## 3. Temporal Filtering

The final modeling dataset covered:

- 2010 Q1 → 2025 Q4

This reduced missing values and ensured consistent availability across all features.

---

# Feature Engineering

## Original Time-Series Features

The following autoregressive forecasting features were retained:

- lag_1
- lag_2
- lag_3
- lag_4
- rolling_mean_4
- rolling_std_4
- time_index

These capture:
- short-term memory,
- seasonal dynamics,
- volatility,
- long-term trend evolution.

---

## External Economic Features

Added economic indicators:

- labor_persons
- labor_jobs
- labor_hours
- compensation
- wages
- labor_costs

---

## Growth-Based Transformations

To better capture economic dynamics, growth-based features were engineered.

### Quarter-over-Quarter (QoQ)

For each economic variable:
- percentage growth relative to previous quarter

### Year-over-Year (YoY)

For each economic variable:
- percentage growth relative to same quarter previous year

Examples:
- labor_persons_qoq_growth
- wages_yoy_growth
- labor_costs_qoq_growth

---

# Modeling Strategy

## Global ML Models

Two forecasting models were evaluated:

### Random Forest Regressor
### Gradient Boosting Regressor

Using:
- TimeSeriesSplit cross-validation
- expanding-window forecasting
- 5 folds

Evaluation metrics:
- MAE
- RMSE
- MAPE

---

# Results

## Random Forest with Extended Features

| Metric | Score |
|---|---|
| MAE | 0.455 |
| RMSE | 0.599 |
| MAPE | 9.914% |

Performance slightly degraded compared to earlier autoregressive-only models.

---

## Random Forest with Growth Features

| Metric | Score |
|---|---|
| MAE | 0.422 |
| RMSE | 0.559 |
| MAPE | 9.250% |

Growth-based transformations improved performance compared to raw economic features.

---

## Gradient Boosting with Growth Features

| Metric | Score |
|---|---|
| MAE | 0.419 |
| RMSE | 0.546 |
| MAPE | 9.145% |

This became the best-performing global ML configuration in this notebook.

---

# Feature Importance Analysis

## Global Feature Importance

The dominant features remained:

| Feature | Approximate Importance |
|---|---|
| rolling_mean_4 | ~46% |
| lag_3 | ~43% |

This confirmed that:
- absenteeism is strongly autoregressive,
- historical dynamics dominate forecasting behavior.

---

## Economic Features Contribution

External variables contributed modestly.

Most relevant external features included:
- labor_costs_qoq_growth
- labor_persons_yoy_growth
- wages_yoy_growth

However, their influence was much smaller than autoregressive features.

---

# Local Sector Modeling

To investigate sector heterogeneity, local forecasting models were trained separately per sector.

This revealed substantial differences in:
- forecastability,
- feature importance,
- optimal model complexity.

---

# Sector Forecasting Performance

| Sector | MAPE |
|---|---|
| P Onderwijs | 6.0% |
| C Industrie | 6.6% |
| A-U Alle economische activiteiten | 7.0% |
| F Bouwnijverheid | 7.1% |
| O Openbaar bestuur | 7.8% |
| Q Gezondheidszorg | 8.8% |
| K Financiële dienstverlening | 10.0% |
| Agriculture | 10.2% |
| Real Estate | 10.3% |
| ICT | 10.4% |

Key insight:
- some sectors are fundamentally easier to forecast than others.

---

# Sector-Specific Feature Behavior

Local feature importance analysis showed:
- sector-specific economic dynamics,
- heterogeneous forecasting drivers.

Example findings:
- Agriculture heavily depended on `time_index`
- ICT required deeper nonlinear modeling
- Education exhibited highly stable patterns

This demonstrated that:
- global models average away sector-specific dynamics.

---

# Hyperparameter Optimization

Targeted GridSearchCV tuning was performed using:
- TimeSeriesSplit
- MAE optimization

Sectors tuned:
- Education
- ICT
- Healthcare
- Agriculture

---

# Hyperparameter Insights

## Education

Preferred:
- shallow trees,
- simple forecasting structure.

## ICT

Preferred:
- deeper trees,
- more nonlinear complexity.

## Agriculture

Preferred:
- trend-dominated simple structures.

## Healthcare

Preferred:
- stronger smoothing,
- noise reduction.

---

# Key Conclusions

## 1. Absenteeism is strongly autoregressive

Historical absenteeism patterns remain the strongest forecasting signal.

---

## 2. External economic features provide incremental value

Economic growth indicators slightly improved forecasting accuracy, especially after transformation into growth features.

---

## 3. Sector heterogeneity is highly important

Forecastability differs significantly across sectors.

---

## 4. Local models outperform generic global logic

Sector-specialized forecasting models reveal richer dynamics and more meaningful economic relationships.

---

## 5. Different sectors require different model complexity

Optimal hyperparameters varied across sectors, indicating that:
- one universal forecasting architecture is likely suboptimal.

---

# Final Takeaway

This notebook demonstrated that:

- external economic indicators can enhance absenteeism forecasting,
- growth-based feature engineering is more effective than raw economic levels,
- sector-specific forecasting strategies provide substantial analytical value,
- forecasting dynamics vary structurally across industries.

The next step is to consolidate all forecasting approaches into a unified benchmark comparison framework.